shared/applicability/cpes.yml

```diff
@@ -0,0 +1,325 @@
+    id: ol7
+    title: "Oracle Linux 7"
+    check_id: installed_OS_is_ol7_family
+
+  "cpe:/o:oracle:linux:8":
+    id: ol8
+    title: "Oracle Linux 8"
+    check_id: installed_OS_is_ol8_family
+
+  "cpe:/o:opensuse:leap:42.1":
+    id: opensuse-42.1
+    title: "openSUSE Leap 42.1"
+    check_id: installed_OS_is_opensuse_leap42
+
+  "cpe:/o:opensuse:leap:42.2":
+    id: opensuse-42.2
+    title: "openSUSE Leap 42.2"
+    check_id: installed_OS_is_opensuse_leap42
+
+  "cpe:/o:opensuse:leap:42.3":
+    id: opensuse-42.3
+    title: "openSUSE Leap 42.3"
+    check_id: installed_OS_is_opensuse_leap42
+
+  "cpe:/o:opensuse:leap:15.0":
+    id: opensuse-15
+    title: "openSUSE Leap 15.0"
+    check_id: installed_OS_is_opensuse_leap15
+
+  "cpe:/a:pam":
+    id: pam
+    title: "Package pam is installed"
+    check_id: installed_env_has_pam_package
+
+  "cpe:/o:redhat:enterprise_linux_coreos:4":
+    id: rhcos4
+    title: "Red Hat Enterprise Linux CoreOS 4"
+    check_id: installed_OS_is_rhcos4
+
+  "cpe:/o:redhat:enterprise_linux:6":
+    id: rhel6
+    title: "Red Hat Enterprise Linux 6"
+    check_id: installed_OS_is_rhel6
+
+  "cpe:/o:redhat:enterprise_linux:6::client":
+    id: rhel6-client
+    title: "Red Hat Enterprise Linux 6 Client"
+    check_id: installed_OS_is_rhel6
+
+  "cpe:/o:redhat:enterprise_linux:6::computenode":
+    id: rhel6-computenode
+    title: "Red Hat Enterprise Linux 6 ComputeNode"
+    check_id: installed_OS_is_rhel6
+
+  "cpe:/o:redhat:enterprise_linux:7":
+    id: rhel7
+    title: "Red Hat Enterprise Linux 7"
+    check_id: installed_OS_is_rhel7
+
+  "cpe:/o:redhat:enterprise_linux:7::server":
+    id: rhel7-server
+    title: "Red Hat Enterprise Linux 7 Server"
+    check_id: installed_OS_is_rhel7
+
+  "cpe:/o:redhat:enterprise_linux:7::client":
+    id: rhel7-client
+    title: "Red Hat Enterprise Linux 7 Client"
+    check_id: installed_OS_is_rhel7
+
+  "cpe:/o:redhat:enterprise_linux:7::computenode":
+    id: rhel7-computenode
+    title: "Red Hat Enterprise Linux 7 ComputeNode"
+    check_id: installed_OS_is_rhel7
+
+  "cpe:/o:redhat:enterprise_linux:7::workstation":
+    id: rhel7-workstation
+    title: "red hat enterprise linux 7 workstation"
+    check_id: installed_OS_is_rhel7
+
+  "cpe:/o:redhat:enterprise_linux:8":
+    id: rhel8
+    title: "Red Hat Enterprise Linux 8"
+    check_id: installed_OS_is_rhel8
+
+  "cpe:/o:redhat:enterprise_linux:8::hypervisor":
+    id: rhel8-host
+    title: "Red Hat Virtualization 4 Host"
+    check_id: installed_OS_is_rhv4
+
+  "cpe:/a:redhat:openstack:10":
+    id: rhosp10
+    title: "Red Hat OpenStack Platform 10"
+    check_id: installed_app_is_rhosp10
+
+  "cpe:/a:redhat:openstack:13":
+    id: rhosp13
+    title: "Red Hat OpenStack Platform 13"
+    check_id: installed_app_is_rhosp13
+
+  "cpe:/a:redhat:enterprise_virtualization_manager:4":
+    id: rhvm4
+    title: "Red Hat Virtualization 4 Manager"
+    check_id: installed_app_is_rhv4
```

Comment:
Overall structure needs to be cleaner and clearer like so:

```
￼    - id: rhvm4
  ￼    title: "Red Hat Virtualization 4 Manager"
      cpe:  "cpe:/a:redhat:enterprise_virtualization_manager:4"
￼      check_id: installed_app_is_rhv4

Reply:
there also needs to be an additional key called `additional_cpes` or `include_cpes` that maps to `shared/applicability/cpes.yml`

Reply:
> there also needs to be an additional key called `additional_cpes` or `include_cpes` that maps to `shared/applicability/cpes.yml`

Why, if the CPEs are picked up automatically based on the content included in the benchmark?

Reply:
To not only make the content readable and understandable, but also remove the abstraction of the build system. Makes it easy for content authors and everyone to read and understand additional cpe is being added in `product.yml`

Reply:
I find it rather cumbersome to, after selecting a rule in the profile, to have to add any `cpe-item` it uses to the `product.yml`.
I'd expect the build-system to be handle that automatically for me (include the CPE in the dictionary of the product).

I think the readability and understandability  should be solved in another way.
Maybe "compiled" products yamls? Build logs?

Reply:
having the cpe in the product.yml makes it easy for content authors to understand.... and no... expecting the build system to handle something is a developerism and not a content authoring-ism.

Reply:
> having the cpe in the product.yml makes it easy for content authors to understand....

Understand what?

The main point of the PR (making it dynamically generated) is to avoid the need to edit every products' file to add a new CPE every time a new CPE is created. Because otherwise, it fails SCAPVal, and may fail during scan with "unknown CPE".
I don't want to keep the same "add something here" gating-like mechanism for content to be correctly built.
The use case it handles is:
1. I created a new rule and it needs to use a new CPE, which I also just created;
1. The rule is for a component very common in all linux distros, so I have not specified a `prodtype`.
1. I selected the rule in the profiles of products I care;
1. Oh no, suddenly all other linux products that picked this rule up into the benchmark started to fail SCAP validation, :(
1. Now someone needs to add the CPE to all other products.



> expecting the build system to handle something is a developerism

I call this convenience of automation.

> content authoring-ism.

Where can I find more about this?
I'm very interested in helping content authors and their use cases.

Reply:
> ... having the cpe in the product.yml makes it easy for content authors to understand...

Imagine that there is a CPE in `product.yml`, and I think that it got there by mistake, or that the reason why it was introduced is not valid any more. What would I have to do in order to resolve those doubts?

Reply:
Maybe I don't understand where you are heading...

If a CPE entry in a `product.yml` was not trivial, I would expect comment lines, if there weren't any I'd look into the commit that added it.

Reply:
Everything has a cost - if the automation performs some magic behind the content author's back, that can have a cost, especially when the magic doesn't work right, or if the author would like something slightly different.

Conversely, if there is no automation, the obvious cost is that you have to add config items s.a. CPEs manually. But that's not where it ends - you get risk of copy-paste errors, and you also have to deal with removal of items that are no longer needed or that have been removed by refactoring.

I think I understand @redhatrises concern of automating too much, and I think that we should look both ways - what are the risks of automating too little?
So far it seems that we should require non-trivial manually added CPEs to be described by comments.

Reply:
I agree that automation can make it hard to do something slightly different. But that was not the original point of the question, and I think it can be handled in a different way.

>  So far it seems that we should require non-trivial manually added CPEs to be described by comments. 

Now I'm not sure if your previous question was about:
- the manually added CPE for the product in `product.yml` or
- the automatically added CPEs in the product's `cpe-dictionary`.

To clarify, the automation doesn't add anything to `product.yml` file, it will include `cpe-item` in the `cpe-dictionary`, as the files removed by this PR (rhel8/cpe/rhel8-cpe-dictionary.xml).
And to check the CPEs added by the build system one can check the built dictionary. It should have the same complexity as reviewing the old static files.

Reply:
My question was about items that would have to be added to `product.yml` if there is no automation. The automation is supposed to make sure that only and only CPEs that are needed get added. Conversely, when a CPE is added manually, it has to be kept in sync with the actual content.

You have answered the question, thanks.